crates/compiler/solve/src/specialize.rs

```diff
@@ -780,10 +784,37 @@ fn get_specialization_lambda_set_ambient_function<P: Phase>(
             //
             // THEORY: if something can become an immediate, it will always be available in the
             // local ability store, because the transformation is local (?)
+            //
+            // TODO: I actually think we can get what we need here by examining `derived_env.exposed_types`,
+            // since immediates can only refer to builtins - and in userspace, all builtin types
+            // are available in `exposed_types`.
             let immediate_lambda_set_at_region =
                 phase.get_and_copy_ability_member_ambient_function(imm, lset_region, subs);
 
             Ok(immediate_lambda_set_at_region)
         }
+
+        SpecializationTypeKey::SingleLambdaSetImmediate(imm) => {
+            let module_id = imm.module_id();
+            debug_assert!(module_id.is_builtin());
+
+            let module_types = &derived_env
+                .exposed_types
+                .get(&module_id)
+                .unwrap()
+                .exposed_types_storage_subs;
+
+            // Since this immediate has only one lambda set, the region must be pointing to 1, and
+            // moreover the imported function type is the ambient function of the single lset.
+            debug_assert_eq!(lset_region, 1);
+            let storage_var = module_types.stored_vars_by_symbol.get(&imm).unwrap();
+            let imported = module_types
+                .storage_subs
+                .export_variable_to(subs, *storage_var);
+
+            roc_types::subs::instantiate_rigids(subs, imported.variable);
+
+            Ok(imported.variable)
+        }
```

Comment:
Super excited about this optimization, lol. It's negligible in practice but it makes the mental model so, so much simpler. Good thing we did Encode/Decode first, which are much more complicated, and require a lot of lambda set trickery!